src/Features/Core/Portable/DocumentHighlighting/AbstractDocumentHighlightsService.cs

```diff
@@ -14,13 +14,48 @@
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
 
-namespace Microsoft.CodeAnalysis.Editor.Implementation.ReferenceHighlighting
+namespace Microsoft.CodeAnalysis.DocumentHighlighting
 {
-    internal abstract class AbstractDocumentHighlightsService : IDocumentHighlightsService
+    internal abstract partial class AbstractDocumentHighlightsService : IDocumentHighlightsService
     {
         public async Task<ImmutableArray<DocumentHighlights>> GetDocumentHighlightsAsync(
             Document document, int position, IImmutableSet<Document> documentsToSearch, CancellationToken cancellationToken)
         {
+            var (succeeded, highlights) = await GetDocumentHighlightsInRemoteProcessAsync(
+                document, position, documentsToSearch, cancellationToken).ConfigureAwait(false);
+
+            if (succeeded)
+            {
+                return highlights;
+            }
+
+            return await GetDocumentHighlightsInCurrentProcessAsync(
+                document, position, documentsToSearch, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<(bool succeeded, ImmutableArray<DocumentHighlights> highlights)> GetDocumentHighlightsInRemoteProcessAsync(
+            Document document, int position, IImmutableSet<Document> documentsToSearch, CancellationToken cancellationToken)
+        {
+            using (var session = await TryGetRemoteSessionAsync(
```

Comment:
:bulb: Consider using the `RemoteHostClient.RunOnRemoteHostAsync<T>` extension method since it manages session lifetime for you.